File: CodingExcercises/C++ver/twoLargeNumber.cpp
#include <iostream>
#include <string>
#include <algorithm>

using namespace std;

string findSum(string num1, string num2)
{
    if (num1.empty() && num2.empty())
    {
        return "0";
    }

    if (num1.length() > num2.length())
    {
        swap(num1, num2);
    }

    int n1 = num1.length();
    int n2 = num2.length();
    int diff = n2 - n1;

    string result = "";
    int carry = 0;

    for (int i = n1 - 1; i >= 0; i--)
    {
        int sum = (num1[i] - '0') + (num2[i + diff] - '0') + carry;
        result.push_back(sum % 10 + '0');
        carry = sum / 10;
    }

    for (int i = diff - 1; i >= 0; i--)
    {
        int sum = (num2[i] - '0') + carry;
        result.push_back(sum % 10 + '0');
        carry = sum / 10;
    }

    if (carry)
        result.push_back(carry + '0');
    reverse(result.begin(), result.end());

    return result;
}

int main()
{
    string num1 = "123456789";
    string num2 = "987654321";
    cout << findSum(num1, num2);
    return 0;
}

File: CodingExcercises/PHPver/twoLargeNumber.php
<?php
function findSum($str1, $str2)
{
    // $str1 = strval($num1);
    // $str2 = strval($num2);

    if (strlen($str1) > strlen($str2)) {
        [$str1, $str2] = [$str2, $str1];
    }
    $result = "";
    $n1 = strlen($str1);
    $n2 = strlen($str2);
    $str1 = strrev($str1);
    $str2 = strrev($str2);
    $carry = 0;
    for ($i = 0; $i < $n1; $i++) {
        $sum = ((ord($str1[$i]) - 48) + (ord($str2[$i]) - 48) + $carry);
        $result .= chr($sum % 10 + 48);
        $carry = $sum / 10;
    }
    for ($i = $n1; $i < $n2; $i++) {
        $sum = (ord($str2[$i]) - 48) + $carry;
        $result .= chr($sum % 10 + 48);
        $carry = $sum / 10;
    }

    if ($carry >= 1)
        $result .= chr($carry % 10 + 48);

    $result = strrev($result);

    return $result;
}

echo findSum("6454513248798645648798465132185764623185", "31651644564551321131884656456546542121");
echo '<br>';
echo findSum("288", "33");
echo '<br>';
